### ISSUE TITLE
Running TPCH queries in Acero through Substrait 

### ISSUE THREAD

**Issue by kabicm** [apache/arrow#40614]
### Describe the usage question you have. Please include as many useful details as  possible.

It would be nice if Acero could run the tpc-h queries through substrait, which seems not possible at the moment.

I converted the TPCH Q1 from sql to substrait using DuckDB and then tried running it with Acero. This resulted in the following errors:
- `Invalid: substrait::Expression::Cast::FailureBehavior unspecified; must be FAILURE_BEHAVIOR_RETURN_NULL or FAILURE_BEHAVIOR_THROW_EXCEPTION`: this can be fixed by setting the `FailureBehavior` to `FAILURE_BEHAVIOR_THROW_EXCEPTION` inside the cast relations in substrait.
- `NotImplemented: Unsupported aggregation phase 'AGGREGATION_PHASE_UNSPECIFIED'.  Only INITIAL_TO_RESULT is supported`: this can be fixed by setting the phase to the `INITIAL_TO_RESULT` value. 
- `Invalid: Expected an aggregate call #sum to have a direct reference`: I don't know how to fix this.

Do you have the tpch queries written in substrait in such a way that Acero can run them? Is the Acero's support for substrait enough to run the tpch queries?

Thank you!

### Component(s)

Benchmarking, C++

**Comment by ianmcook** [comment]
@richtia do we have all the TPC-H queries in Substrait format? IIRC there are tests that run these against Acero. Can you give @kabicm a pointer to that? Thanks

**Comment by richtia** [comment]
> @richtia do we have all the TPC-H queries in Substrait format? IIRC there are tests that run these against Acero. Can you give @kabicm a pointer to that? Thanks

Yeah, here are all the TPC-H queries in substrait format (there are actually a few missing because substrait couldn't convert them at the time, need to see if i that's been fixed): https://github.com/substrait-io/consumer-testing/tree/main/substrait_consumer/tests/integration/queries/tpch_substrait_plans

This is the acero test that runs all of them: https://github.com/substrait-io/consumer-testing/blob/main/substrait_consumer/tests/integration/test_acero_tpch.py

Unfortunately Acero isn't at a point where it's able to complete run any of these fully through.

They all fail for various reasons, which can be seen from running the tests.

**Comment by kabicm** [comment]
Thanks @ianmcook and @richtia! 

Great to see the substrait plans for the TPC-H queries, thanks for the pointers! However, I am still getting the same type of issues as mentioned before when trying to run them with Acero.

At the moment, it seems to me that the only way to run the TPCH queries with Acero in C++ is by creating Acero's exec plan. 

Are there maybe Acero query plans for TPC-H queries in C++? I found the Q1 and Q6 Acero plans in C++, using the Declaration API.

**Comment by richtia** [comment]
@kabicm, yeah...Acero isn't at the point where it can fully run any of those TPC-H substrait plans.  Most  of them either hit the issue you've been encountering or this one: 
```
pyarrow.lib.ArrowInvalid: A join rel's expression must be a simple equality between keys but got true
```

From the C++ side, maybe @vibhatha could help you out.  He had been working on a bit of the Acero substrait consumer development previoiusly.

**Comment by vibhatha** [comment]
@kabicm at the moment, I am not doing active developments in Acero/Substrait components. I haven't been contributing to that part of the codebase for a while, though, I can help out with some pointers if improvements are made via PRs. Would that be helpful?

**Comment by kabicm** [comment]
Thanks @richtia and @vibhatha for your fast response and the pointers! I will be keeping an eye on the substrait frontend in Acero.

Sure, @vibhatha, pointers are always welcome! But I think this issue can now be closed, as you are well aware of all the issues with Acero+Substrait, which I mentioned.

**Comment by thisisnic** [comment]
This issue has been marked as stale because it has had no activity in the past 365 days. Please remove the stale label or comment below, or this issue will be closed in 14 days. If this usage question has evolved into a feature request or docs update, please remove the 'Type: usage' label and add the 'Type: enhancement' label instead.
